File: lab_suite/widgets/gain_control.py
"""
GainControlVue – Gain-Control als NiceGUI Custom-Element (Vue/JS).

Verwendung wie die Python-Klasse GainControl:
  from lab_suite.widgets import GainControlVue
  gain = GainControlVue('Gain', min_=0, max_=10, value=1, on_change=lambda e: ...)
  # Wert aus Event: e.args (float)
  # Reset von außen: gain.reset()
"""
from __future__ import annotations

from collections.abc import Callable

from nicegui.element import Element


class GainControlVue(Element, component="gain_control.js"):

    def __init__(
        self,
        label: str = "Gain",
        *,
        min_: float = 0,
        max_: float = 10,
        value: float = 1,
        on_change: Callable | None = None,
    ) -> None:
        super().__init__()
        self._props["label"] = label
        self._props["min"] = min_
        self._props["max"] = max_
        self._props["value"] = value
        if on_change is not None:
            self.on("change", on_change)

    def reset(self) -> None:
        """Reset auf 1 (ruft Methode in der Vue-Komponente auf)."""
        self.run_method("reset")

File: lab_suite/widgets/gain_control.js
// GainControl – Vue-Komponente für NiceGUI (lab_suite/widgets)
// EINZIGE Quelle: lab_suite/widgets/gain_control.js (keine Duplikate).
// Emittiert NUR in: onInput (beim Ziehen) und reset() (Button). Natives @change (Maus loslassen) wird mit .stop abgefangen, damit NiceGUI kein 1.0 nachschiebt.
// Einmal nach Änderung: Browser mit deaktiviertem Cache neu laden.
export default {
  template: `
    <div class="gain-control" style="background:#ccc; padding:8px; border-radius:6px; display:inline-flex; align-items:center; gap:8px;">
      <span>{{ label }}</span>
      <span>{{ displayValue }}</span>
      <input
        type="range"
        :min="min"
        :max="max"
        :value="internalValue"
        step="0.01"
        style="min-width:80px;"
        @input="onInput"
        @change.stop="onNativeChange"
      />
      <button type="button" @click="reset">Reset</button>
    </div>
  `,
  props: {
    label: { type: String, default: 'Gain' },
    min: { type: Number, default: 0 },
    max: { type: Number, default: 10 },
    value: { type: Number, default: 1 },
  },
  data() {
    return {
      internalValue: this.value,
    };
  },
  computed: {
    displayValue() {
      return this.internalValue.toFixed(2);
    },
  },
  watch: {
    value(v) {
      const num = Number(v);
      if (Number.isNaN(num)) return;
      if (num === 1 && this.internalValue !== 1) return;
      this.internalValue = num;
    },
  },
  methods: {
    onInput(ev) {
      const v = parseFloat(ev.target.value);
      if (Number.isNaN(v)) return;
      this.internalValue = v;
      this._emitValue(v);
    },
    onNativeChange(ev) {
      // Natives change (Maus loslassen) stoppen – sonst leitet NiceGUI es weiter und sendet 1.0
      ev.stopPropagation();
      ev.preventDefault();
    },
    reset() {
      this.internalValue = 1;
      this._emitValue(1);
    },
    _emitValue(v) {
      const num = Number(v);
      if (!Number.isNaN(num)) this.$emit('change', num);
    },
  },
};
